flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointBriefCalculator.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.Execution;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Computes the tasks to trigger, wait or commit for each checkpoint. */
+public class CheckpointBriefCalculator {
+    private static final Logger LOG = LoggerFactory.getLogger(CheckpointBriefCalculator.class);
+
+    private final JobID jobId;
+
+    private final List<ExecutionVertex> tasksToTrigger;
+
+    private final List<ExecutionVertex> tasksToWait;
+
+    private final List<ExecutionVertex> tasksToCommitTo;
+
+    public CheckpointBriefCalculator(
+            JobID jobId,
+            List<ExecutionVertex> tasksToTrigger,
+            List<ExecutionVertex> tasksToWait,
+            List<ExecutionVertex> tasksToCommitTo) {
+
+        this.jobId = jobId;
+        this.tasksToTrigger = Collections.unmodifiableList(tasksToTrigger);
+        this.tasksToWait = Collections.unmodifiableList(tasksToWait);
+        this.tasksToCommitTo = Collections.unmodifiableList(tasksToCommitTo);
+    }
+
+    public CheckpointBrief calculateCheckpointBrief() throws CheckpointException {
+        List<Execution> tasksToTrigger = getTriggerExecutions();
+        Map<ExecutionAttemptID, ExecutionVertex> tasksToWait = getAckTasks();
+
+        return new CheckpointBrief(
+                Collections.unmodifiableList(tasksToTrigger), tasksToWait, tasksToCommitTo);
```

Comment:
We could not mark `tasksToWait` as readonly since in `PendingCheckpoint` it would be reused to record not-acknowledged tasks and would be modified if tasks report snapshot. If we change it to read-only, we would need to clone it when creating `PendingCheckpoint`.

Reply:
I think we should do that (make it unmodifiable here and copy in `PendingCheckpoint`). I expect `CheckpointBrief` to be immutable.

Besides, this class isn't concerned about mutability of these three structures: two are newly created, one is already immutable (`tasksToCommitTo`).
So I'd move wrapping with `unmodifiable` to the `CheckpointBrief` constructor.

nit: inline variables?

Reply:
I also agree with that make it unmodifiable would be more clean and more error-prone. Thus I made it unmodifiable and copy in `PendingCheckpoint`. 

I also make the variables inline.